src/System.Linq.Expressions/src/System/Linq/Expressions/BlockExpression.cs

```diff
@@ -937,5 +955,20 @@ internal static void ValidateVariables(ReadOnlyCollection<ParameterExpression> v
                 set.Add(v);
             }
         }
+
+        private static BlockExpression GetOptimizedBlockExpression(IReadOnlyList<Expression> expressions)
+        {
+            switch (expressions.Count)
+            {
+                case 2: return Block(expressions[0], expressions[1]);
+                case 3: return Block(expressions[0], expressions[1], expressions[2]);
+                case 4: return Block(expressions[0], expressions[1], expressions[2], expressions[3]);
+                case 5: return Block(expressions[0], expressions[1], expressions[2], expressions[3], expressions[4]);
+                default:
+                    ContractUtils.RequiresNotEmptyList(expressions, "expressions");
+                    RequiresCanRead(expressions, "expressions");
+                    return new BlockN(expressions.ToArray());
```

Comment:
It will be hit if `Expressions` is accessed, which will definitely happen if it gets rewritten or visited, which is one of the reasons why the method affected by this change might be hit rather than the `params` method. Since `ToReadOnly` returns the argument if its already a `ReadOnlyCollection` and uses a single instance for all empty collections of the same element type, the consistent use of it throughout means there's a high chance of instances being shared, which is hence lighter than using an array.

A middle ground would be to use:

```C#
return new BlockN(expressions as ReadOnlyCollection<Expression> ?? (IList<Expression>)expressions.ToArray());
```

As then the method that takes an `IEnumerable<Expression>` and which is hit in the instance-sharing cases will still have the previous behaviour of using the `ReadOnlyCollection` it passes and, the method with the single `params` methods will keep its previous behaviour of passing a new copy of the array, which as you say is lighter if `Expressions` is never hit.